### PR TITLE
libbpf-cargo/btf: Add support for BTF_KIND_ENUM64

### LINKED ISSUE #307: Add `enum64` support
We should add support for `enum64`, as it has been added [upstream](https://lore.kernel.org/bpf/165462361769.5218.5561986964163226721.git-patchwork-notify@kernel.org/) a while back.

### PR DESCRIPTION
**Pull request by chenhengqi**
Closes #307.

### CODE REVIEW

File: libbpf-cargo/src/btf/c_types.rs [hunk 1]
```diff
@@ -32,6 +32,14 @@ pub struct btf_enum {
     pub val: i32,
 }
 
+#[repr(C)]
+#[derive(Debug, Clone, DerivePread, Pwrite, IOread, SizeWith)]
+pub struct btf_enum64 {
```
Comment (danielocfb) [inline libbpf-cargo/src/btf/c_types.rs:37]: Can we use the [`libbpf_sys` definition](https://docs.rs/libbpf-sys/1.1.1+v1.1.0/libbpf_sys/struct.btf_enum64.html) instead?

File: libbpf-cargo/src/btf/btf.rs [hunk 2]
```diff
@@ -993,6 +1038,49 @@ impl Btf {
                         writeln!(def, r#"}}"#)?;
                     }
                 }
+                BtfType::Enum64(t) => {
```
Comment (danielocfb) [inline libbpf-cargo/src/btf/btf.rs:1041]: This is fine in principle, but please let's start introducing functions for newly added branches:
```patch
--- libbpf-cargo/src/btf/btf.rs
+++ libbpf-cargo/src/btf/btf.rs
@@ -765,6 +765,52 @@ impl Btf {
         }
     }

+
+    fn enum64_def(btf: &BtfEnum64<'_>, def: &mut String) -> Result<()> {
+        let repr_size = match btf.size {
+            1 => "8",
+            2 => "16",
+            4 => "32",
+            8 => "64",
+            _ => bail!("Invalid enum64 size: {}", btf.size),
+        };
+
+        let signed = if btf.signed { "i" } else { "u" };
+
+        writeln!(def, r#"#[derive(Debug, Copy, Clone, PartialEq, Eq)]"#)?;
+        writeln!(def, r#"#[repr({signed}{repr_size})]"#)?;
+        writeln!(def, r#"pub enum {name} {{"#, name = btf.name)?;
+
+        for value in &btf.values {
+            writeln!(
+                def,
+                r#"    {name} = {value},"#,
+                name = value.name,
+                value = value.value,
+            )?;
+        }
+
+        writeln!(def, "}}")?;
+
+        // write an impl Default for this enum
+        if !btf.values.is_empty() {
+            // TODO: remove #[allow(clippy::derivable_impls)]
+            //       once minimum rust at 1.62+
+            writeln!(def, r#"#[allow(clippy::derivable_impls)]"#)?;
+            writeln!(def, r#"impl Default for {name} {{"#, name = btf.name)?;
+            writeln!(def, r#"    fn default() -> Self {{"#)?;
+            writeln!(
+                def,
+                r#"        {name}::{value}"#,
+                name = btf.name,
+                value = btf.values[0].name
+            )?;
+            writeln!(def, r#"    }}"#)?;
+            writeln!(def, r#"}}"#)?;
+        }
+        Ok(())
+    }
+
     /// Returns rust type definition of `ty` in string format, including dependent types.
     ///
     /// `ty` must be a struct, union, enum, or datasec type.
@@ -1038,49 +1084,7 @@ impl Btf {
                         writeln!(def, r#"}}"#)?;
                     }
                 }
-                BtfType::Enum64(t) => {
-                    let repr_size = match t.size {
-                        1 => "8",
-                        2 => "16",
-                        4 => "32",
-                        8 => "64",
-                        _ => bail!("Invalid enum64 size: {}", t.size),
-                    };
-
-                    let signed = if t.signed { "i" } else { "u" };
-
-                    writeln!(def, r#"#[derive(Debug, Copy, Clone, PartialEq, Eq)]"#)?;
-                    writeln!(def, r#"#[repr({signed}{repr_size})]"#)?;
-                    writeln!(def, r#"pub enum {name} {{"#, name = t.name)?;
-
-                    for value in &t.values {
-                        writeln!(
-                            def,
-                            r#"    {name} = {value},"#,
-                            name = value.name,
-                            value = value.value,
-                        )?;
-                    }
-
-                    writeln!(def, "}}")?;
-
-                    // write an impl Default for this enum
-                    if !t.values.is_empty() {
-                        // TODO: remove #[allow(clippy::derivable_impls)]
-                        //       once minimum rust at 1.62+
-                        writeln!(def, r#"#[allow(clippy::derivable_impls)]"#)?;
-                        writeln!(def, r#"impl Default for {name} {{"#, name = t.name)?;
-                        writeln!(def, r#"    fn default() -> Self {{"#)?;
-                        writeln!(
-                            def,
-                            r#"        {name}::{value}"#,
-                            name = t.name,
-                            value = t.values[0].name
-                        )?;
-                        writeln!(def, r#"    }}"#)?;
-                        writeln!(def, r#"}}"#)?;
-                    }
-                }
+                BtfType::Enum64(t) => Self::enum64_def(t, &mut def)?,
                 BtfType::Datasec(t) => {
                     let mut sec_name = t.name.to_string();
                     if sec_name.is_empty() || !sec_name.starts_with('.') {
```

File: libbpf-cargo/src/btf/types.rs [hunk 3]
```diff
@@ -183,6 +184,20 @@ pub struct BtfTypeTag<'a> {
...
 
+#[derive(Debug)]
+pub struct BtfEnum64Value<'a> {
+    pub name: &'a str,
+    pub value: u64,
+}
+
+#[derive(Debug)]
+pub struct BtfEnum64<'a> {
+    pub name: String,
```
Comment (danielocfb) [inline libbpf-cargo/src/btf/types.rs:195]: Why do we need an owned string here, as opposed to `&'a str`?

File: libbpf-cargo/src/btf/c_types.rs (hunk 1 above)
Comment (chenhengqi) [reply libbpf-cargo/src/btf/c_types.rs:37]: libbpf_sys::btf_enum64 does NOT derive `DerivePread`.

File: libbpf-cargo/src/btf/types.rs (hunk 3 above)
Comment (chenhengqi) [reply libbpf-cargo/src/btf/types.rs:195]: Enum could be anonymous, so we name it.

```rust
let name = match Self::get_btf_str(string_table, t.name_off as usize)? {
    "" => {
        self.anon_count += 1;
        format!("{}{}", ANON_PREFIX, self.anon_count)
    }
    n => n.to_string(),
};
```

**Comment by chenhengqi** [comment]
Don't think the CI supports Enum64, will try.

**Comment by danielocfb** [comment]
> Don't think the CI supports Enum64, will try.

Thanks! If it's not supported but you can at least try it locally, let's just mark it `#[ignored]` for now so that we can enable it later on.

File: libbpf-cargo/src/btf/c_types.rs (hunk 1 above)
Comment (danielocfb) [reply libbpf-cargo/src/btf/c_types.rs:37]: I missed that. That's rather unfortunate :-|

**Comment by danielocfb** [comment]
With the rewrite of the btf logic this change will need a bit of a rework, unfortunately. @chenhengqi are you still interested in doing that?

**Comment by chenhengqi** [comment]
Didn't quite follow here.

**Pull request closed without merging**